### PR TITLE
Add `prettyprint` (part 1/3 refactor text representation of objects)

### PR DESCRIPTION
**Pull request by t-imamichi**
### Summary

This is part 1 of #184.

This PR introduces `prettyprint` function and `QuadraticProgram.prettyprint` method to generate a pretty-printed string of a `QuadraticProgram` object.

### Details and comments

We use Docplex's `export_as_lp_string` as a kind of pretty print of QP now.
But, it's good to have our own pretty printing format that is more comprehensive than Docplex's.
Moreover, it makes us easier to deal with pretty printing of planned new features such as higher order polynomials and semi-definite programming (Docplex cannot handle them).

The format is based on the discussion https://github.com/Qiskit/qiskit-optimization/pull/184#issuecomment-899025806
~~A minor modification is that I omitted a symbol `*` of the multiplication of two variables, e.g., `x * y` (before) -> `x y` (current) because the discussion did not have `*` of the multiplication of a scalar coefficient and a variable, e.g., `2 x` (before and current). So, I make them consistent. The output of the multiplication of a scalar and two variables would be `2 x * y` (before) -> `2 x y` (current)~~

I updated the multiplication to use only `*` between scalars and variables.

Variable names with " ", "+", "-", and "*" are translated into `("{name}")` to avoid confusion. These symbols are used to represent expressions.

Minor change:
- Updated `QuadraticProgram` to show a warning message if non-printable strings are set as problem name, variable name, or constraint name.
  - `prettyprint` raises an error if non-printable string is given, because it cannot be displayed.
- Updated `QuadraticProgram.{binary_var, integer_var, continuous_var}` to use the default name if an empty name is given.

Example 1: usual use case
```python
from qiskit_optimization.problems import QuadraticProgram
from qiskit_optimization.translators.prettyprint import prettyprint

qp = QuadraticProgram('problem 1')
qp.integer_var(-1, 2, 'x')
qp.integer_var(-1, 2, 'y')
qp.continuous_var(-1, name='z')
qp.binary_var('u')
qp.binary_var('v')
qp.binary_var_list(10)
qp.integer_var_list(3,name='φ')
qp.continuous_var_list(3,name='η')
qp.minimize(constant=3, linear={'x': 2, 'y': 3}, quadratic={('u', 'x'): -1})
qp.linear_constraint({'x': 1, 'y': -2}, '>=', 2, name='lin_GE')
qp.linear_constraint({'x': 2, 'y': -1}, '==', 1, name='lin_EQ')
qp.quadratic_constraint({'x': 1, 'u': 1}, {(3, 4): 1, (5, 6): -1}, '<=', 1, name='quad_LE')
qp.quadratic_constraint({'x': 2, 'y': -1}, {('z', 'z'): -1}, '<=', 1)

# the following two lines return the same output
#print(prettyprint(qp))
print(qp.prettyprint())
```

output
```
Problem name: problem 1

Minimize
  -x*u + 2*x + 3*y + 3

Subject to
  Linear constraints (2)
    x - 2*y >= 2  'lin_GE'
    2*x - y == 1  'lin_EQ'

  Quadratic constraints (2)
    u*v - x5*x6 + u + x <= 1  'quad_LE'
    -z^2 + 2*x - y <= 1  'q1'

  Integer variables (5)
    -1 <= x <= 2
    -1 <= y <= 2
    0 <= φ15
    0 <= φ16
    0 <= φ17

  Continuous variables (4)
    -1 <= z
    0 <= η18
    0 <= η19
    0 <= η20

  Binary variables (12)
    u v x5 x6 x7 x8 x9 x10 x11 x12 x13 x14
```

------------------
Example 2: corner case of variable names with symbols
The following does not take a special care of `+u` or `-v`; it is a similar way of docplex's prettyprint and gurobipy's display. https://github.com/Qiskit/qiskit-optimization/pull/366#issuecomment-1111800754
```python
from qiskit_optimization.problems import QuadraticProgram

qp = QuadraticProgram('problem 1')
qp.integer_var(-1, 2, 'x')
qp.integer_var(-1, 2, 'y')
qp.continuous_var(-1, name=' z')
qp.binary_var('+u')
qp.binary_var('-v')
qp.binary_var('*w')
qp.binary_var_list(10)
qp.integer_var_list(3, name='φ')
qp.continuous_var_list(3, name='η')
qp.minimize(constant=3, linear={'x': 2, 'y': 3}, quadratic={('+u', 'x'): -1})
qp.linear_constraint({'x': 1, 'y': -2}, '>=', 2, name='lin_GE')
qp.linear_constraint({'x': 2, 'y': -1}, '==', 1, name='lin_EQ')
qp.quadratic_constraint({'x': 1, '+u': 1}, {(3, 4): 1, (5, 6): -1}, '<=', 1, name='quad_LE')
qp.quadratic_constraint({'x': 2, 'y': -1}, {(' z', ' z'): -1}, '<=', 1)

print(qp.prettyprint())
```

output
```
Problem name: problem 1

Minimize
  -x*+u + 2*x + 3*y + 3

Subject to
  Linear constraints (2)
    x - 2*y >= 2  'lin_GE'
    2*x - y == 1  'lin_EQ'

  Quadratic constraints (2)
    -*w*x6 + +u*-v + +u + x <= 1  'quad_LE'
    - z^2 + 2*x - y <= 1  'q1'

  Integer variables (5)
    -1 <= x <= 2
    -1 <= y <= 2
    0 <= φ16
    0 <= φ17
    0 <= φ18

  Continuous variables (4)
    -1 <=  z
    0 <= η19
    0 <= η20
    0 <= η21

  Binary variables (13)
    +u -v *w x6 x7 x8 x9 x10 x11 x12 x13 x14 x15
```

### CODE REVIEW

**Comment by woodsp-ibm** [comment]
I must admit I find having spaces there, instead of *, makes it that I need to look more closely at things. Its not so easy to see at a glance. The spaces can be interpreted either as separation or multiply and that duality, for me, makes the one with spaces take more effort to look at (versus one with explicit *) as per below
```
u v - x5 x6 + u + x <= 1

u*v - x5*x6 + u + x <= 1
```
I see that docplex leaves out the multiply, as you say, on constants, but why not have it, and then have consistency with it present in all places.
```
x - 2 y >= 2

x - 2*y >= 2
```

I think removing the space, or multiply, and simply putting the variables together
```
uv  - x5x6 + u + x <= 1

x - 2y >= 2
```
while that might work it could become ambiguous depending on the user choice of variables should I also have a variable 'uv' or a variable '2y' (I don't think there are any naming restrictions on doing that right). So that seems like a non-starter.

**Comment by t-imamichi** [comment]
Thank you for the feedback on the appearance. I want a format that is easy to see at a glance.
I think we need at least one character to separate variables because variable name "uv" might be possible. 
I'm fine to use `"*"` for multiplication. What do you think of the following examples?
```
option 1: 2 u v - 3 x5 x6 + 4 u - x <= 1 (current)

option 2: 2*u*v - 3*x5*x6 + 4*u - x <= 1 (use "*" without space)

option 3: 2 * u * v - 3 * x5 * x6 + 4 * u - x <= 1 (use "*" with space)
```
I notice that the output would be complicated If users use `"*"` or `" "` as part of variable names. Perhaps it's good to warn it when `prettyprint` recognizes `"*"` and `" "` in the variable names.

**Comment by woodsp-ibm** [comment]
For me personally, of what you have above option 2 seems easier to look at. How about adding ( ) to that as below...

```
2 * u * v - 3 * x5 * x6 + 4 * u - x <= 1

(2 * u * v) - (3 * x5 * x6) + (4 * u) - x <= 1
```
For option 3 with all spaces the grouping does not stand out so easily. One could add brackets - makes it a bit more verbose but for me its easier to look at. So that or option 2, though this would help if they have * in the name; though I guess if you really want to cause yourself problems with the output you can have a variable like '-x' as well as one called 'x' then good luck working out what '-x' is in the output. I guess the first one negated would be '--x'!

**Comment by t-imamichi** [comment]
Thank you for your idea. I prefer a short format because these strings will be used in debuggers of PyCharm and vscode.
I tried the option 2 in the latest commit and updated the description. https://github.com/Qiskit/qiskit-optimization/pull/366#issue-1214171077
As for problematic symbols (`" ", "-", "+", "*"), which are used in pretty-printed expressions, I tried to display them with `("{name}")`. See the example of the description.

I see Docplex rename such names with the symbols into a different names with a warning message. I think it might be confusing to users. So, I didn't try this approach.
```python
from docplex.mp.model import Model

mod = Model()
mod.binary_var(' x')
mod.binary_var('*x')
mod.binary_var('/x')
mod.binary_var('+x')
mod.binary_var('-x')

print(mod.export_as_lp_string())
```
output
```
Warning: Some identifiers are not valid LP identifiers: 1 (e.g.: "*x")
\ This file has been generated by DOcplex
\ ENCODING=ISO-8859-1
\Problem name: docplex_model1

Minimize
 obj:
Subject To

Bounds
 0 <= _x <= 1
 0 <= x2 <= 1
 0 <= _x#2 <= 1
 0 <= px <= 1
 0 <= mx <= 1

Binaries
 _x x2 _x#2 px mx
End
```

**Comment by woodsp-ibm** [comment]
>  I prefer a short format because these strings will be used in debuggers of PyCharm and vscode.

This here though is some more user friendly pretty printed string right - more so than repr or perhaps str. Maybe there will be similarities hence the comment.

I agrees this `"*a*"*"*b*"` is hard on the eyes but does it need to be wrapped with both ( and " ie `("*a*")*("*b*")` which makes it longer but I agree a bit more readable. Maybe it doesn't matter too much as hopefully the user would avoid using those characters in the name anyway so what we are talking about is more of an edge case. We could choose to do nothing special and just note that a poor choice of names, ones that include spaces etc may lead to various string outputs being harder to parse.

In looking at LP format they do have a definition of what a variable name can be https://www.ibm.com/docs/en/icos/12.8.0.0?topic=representation-variable-names-in-lp-file-format hence I guess docplex cannot legitimately put them into the LP output.

**Comment by t-imamichi** [comment]
I tried the corner case with docplex and gurobipy. Docplex has `prettyprint` and `export_as_lp_string`.

Here is a summary of a sum of various corner cases.
```
docplex - prettyprint:  x^3 + 2x + .4 + *x + y%4 + a_m_b + θ + x10 [ 1.2*[5] ];
docplex - lp string:    x1 + x2 + x3 + x6 + y%4 + a_m_b + x9 + x10 + [ 2 x4*x5 ]/2
gurobipy:  <gurobi.QuadExpr: x^3 + 2x + .4 + *x + y+0x1.cp-1072 - b + θ + C9 + [ 1.2 * [5] ]>
```

LP format has various restrictions as you pointed out and it converts many variable names into default one (`x_{}`).
But, pretty-printed format of both docplex and gurobipy transform only some specific corner cases.
I think our pretty-printed format can take a similar approach with docplex and gurobipy, i.e., we leave symbols as is.
But, we have to convert an empty name into a default name.

Btw, gurobipy has a strange output such as `0x1.cp-1072`. They might have an issue with `%`.

-----------------
Details

```python
from docplex.mp.model import Model

mod = Model()

a = mod.binary_var('x^3')
b = mod.binary_var('2x')
c = mod.binary_var('.4')
d = mod.binary_var('1.2')
e = mod.binary_var('[5]')
f = mod.binary_var('*x')
g = mod.binary_var('y%4')
h = mod.binary_var('a - b')
i = mod.binary_var('θ')
j = mod.binary_var('')

mod.minimize(a + b + c + d * e + f + g + h + i +j)

mod.prettyprint()
print('-----------------')
print(mod.export_as_lp_string())
```

output:
```
// This file has been generated by DOcplex
// model name is: docplex_model1
// single vars section
dvar bool x^3;
dvar bool 2x;
dvar bool .4;
dvar bool 1.2;
dvar bool [5];
dvar bool *x;
dvar bool y%4;
dvar bool a_m_b;
dvar bool θ;
dvar bool x10;

minimize
 x^3 + 2x + .4 + *x + y%4 + a_m_b + θ + x10 [ 1.2*[5] ];

subject to {

}
-----------------
Warning: Some identifiers are not valid LP identifiers: 7 (e.g.: "x^3")
\ This file has been generated by DOcplex
\ ENCODING=ISO-8859-1
\Problem name: docplex_model1

Minimize
 obj: x1 + x2 + x3 + x6 + y%4 + a_m_b + x9 + x10 + [ 2 x4*x5 ]/2
Subject To

Bounds
 0 <= x1 <= 1
 0 <= x2 <= 1
 0 <= x3 <= 1
 0 <= x4 <= 1
 0 <= x5 <= 1
 0 <= x6 <= 1
 0 <= y%4 <= 1
 0 <= a_m_b <= 1
 0 <= x9 <= 1
 0 <= x10 <= 1

Binaries
 x1 x2 x3 x4 x5 x6 y%4 a_m_b x9 x10
End
```
-----------------
```python
import gurobipy as gp

mod = gp.Model()

a = mod.addVar(vtype=gp.GRB.BINARY, name='x^3')
b = mod.addVar(vtype=gp.GRB.BINARY, name='2x')
c = mod.addVar(vtype=gp.GRB.BINARY, name='.4')
d = mod.addVar(vtype=gp.GRB.BINARY, name='1.2')
e = mod.addVar(vtype=gp.GRB.BINARY, name='[5]')
f = mod.addVar(vtype=gp.GRB.BINARY, name='*x')
g = mod.addVar(vtype=gp.GRB.BINARY, name='y%4')
h = mod.addVar(vtype=gp.GRB.BINARY, name='a - b')
i = mod.addVar(vtype=gp.GRB.BINARY, name='θ')
j = mod.addVar(vtype=gp.GRB.BINARY, name='')

mod.setObjective(a + b + c + d * e + f + g + h + i + j, gp.GRB.MINIMIZE)
mod.update()
mod.display()
```
output
```
Minimize
  <gurobi.QuadExpr: x^3 + 2x + .4 + *x + y+0x1.cp-1072 - b + θ + C9 + [ 1.2 * [5] ]>
Subject To
Binaries
  ['x^3', '2x', '.4', '1.2', '[5]', '*x', 'y 0x1.ffdefa1fe7p-1028 - b', 'θ', 'C9']
```

**Comment by t-imamichi** [comment]
I updated the PR as follows.
- Remove the special care `("{name}")` similar to docplex.prettyprint and gurobipy.dispaly
- Set the default variable name if an empty name is given

**Comment by t-imamichi** [comment]
The corner case with this PR

```python
from qiskit_optimization.problems import QuadraticProgram

mod = QuadraticProgram()

a = mod.binary_var('x^3')
b = mod.binary_var('2x')
c = mod.binary_var('.4')
d = mod.binary_var('1.2')
e = mod.binary_var('[5]')
f = mod.binary_var('*x')
g = mod.binary_var('y%4')
h = mod.binary_var('a - b')
i = mod.binary_var('θ')
j = mod.binary_var('')

mod.minimize(linear=[1, 1, 1, 0, 0, 1, 1, 1, 1, 1], quadratic={(3, 4): 1})

print(mod.prettyprint())
```

output
```
Problem name: 

Minimize
  1.2*[5] + *x + .4 + 2x + a - b + x9 + x^3 + y%4 + θ

Subject to
  No constraints

  Binary variables (10)
    x^3 2x .4 1.2 [5] *x y%4 a - b θ x9
```

**Comment by t-imamichi** [comment]
I think it's easy for users to avoid such problematic cases (not use white spaces, etc.). So, I'm fine not to include a special care for corner cases similar to docplex.prettyprint and gurobipy.display.

File: qiskit_optimization/problems/quadratic_program.py [hunk 1]
```diff
@@ -147,7 +152,7 @@ def _add_variable(
         vartype: VarType,
         name: Optional[str],
     ) -> Variable:
-        if name is None:
+        if not name:
```
[the remark this answers is not in the document]
Comment (t-imamichi) [reply qiskit_optimization/problems/quadratic_program.py:155]: Yes, that's my intention to replace an empty name with a default name and I wrote a description in reno. Empty variable name is problematic for pretty printing.
This behavior is same as constraint names. If we set an empty name to constraint name, it sets the default name.

File: qiskit_optimization/problems/quadratic_program.py [hunk 2]
```diff
@@ -110,6 +113,8 @@ def name(self, name: str) -> None:
         Args:
             name: The name of the quadratic program.
         """
+        if not name.isprintable():
+            warn("Problem name is not printable")
```
[the remark this answers is not in the document]
Comment (t-imamichi) [reply qiskit_optimization/problems/quadratic_program.py:117]: Thanks. Printing a non-printable name with `repr` is a great idea.

**Comment by t-imamichi** [comment]
I updated the warning and error message to include `repr(name)` and simplify the problem name setting.

File: qiskit_optimization/problems/quadratic_program.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (t-imamichi) [reply qiskit_optimization/problems/quadratic_program.py:155]: FYI: diff of main and this PR.

```python
from qiskit_optimization import QuadraticProgram

qp = QuadraticProgram()
x = qp.binary_var(name='')
c = qp.linear_constraint(name='')
q = qp.quadratic_constraint(name='')
print('variable:', x.name)
print('lin cst:', c.name)
print('quad cst:', q.name)
```

main
```
variable: 
lin cst: c0
quad cst: q0
```

this PR
```
variable: x0
lin cst: c0
quad cst: q0
```

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml [hunk 3]
```diff
@@ -0,0 +1,95 @@
+---
+upgrade:
+  - |
+    Shows a warning message if non-printable strings are set to
+    :class:`~qiskit_optimization.problems.QuadraticProgram`
+    as problem name, variable name, or constraint name.
+  - |
+    If users set an empty variable name ``""`` with
+    :meth:`~qiskit_optimization.problems.QuadraticProgram.binary_var`,
+    :meth:`~qiskit_optimization.problems.QuadraticProgram.integer_var`, and
```
Comment (woodsp-ibm) [inline releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:10]: I wish I could see why these do not seem to be coming out as links in the html. They are highlighted but no clickable as if it failed to resolve them but I cannot see why.

File: qiskit_optimization/problems/quadratic_program.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (woodsp-ibm) [reply qiskit_optimization/problems/quadratic_program.py:155]: I am fine with the the empty name change - you noted it in the reno. It seems to be a change for the better.

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml (hunk 3 above)
Comment (t-imamichi) [reply releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:10]: Yeah, I cannot see the links to `QuadraticProgram` and methods `binary_var`, etc., either.
Do you have any idea, @manoelmarques?

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+features:
+  - |
+    Adds :func:`~qiskit_optimization.translator.prettyprint.prettyprint`
+    to generate a pretty-printed string of
+    :class:`~qiskit_optimization.problem.QuadraticProgram`.
+  - |
+    Adds :meth:`~qiskit_optimization.problem.QuadraticProgram.prettyprint`
+    to generate a pretty-printed string of the object.
+
+    Here is an example of pretty printing.
```
[the remark this answers is not in the document]
Comment (t-imamichi) [reply releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:36]: Thanks. It's a good idea to have only `QP.prettyprint` and remove `prettyprint` function from `__init__` and document.

**Comment by t-imamichi** [comment]
I updated the PR 88f28d9 as follows
- Remove `prettyprint` function in `__all__` of translators
- Mention default name in QP`s methods ("if `name` is None or empty, the default name is used").
- Adds `qp.prettyprint` and the unit test of prettyprint.

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml (hunk 3 above)
Comment (woodsp-ibm) [reply releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:10]: I figured why - we have import the class to the root and document it there ie at `~qiskit_optimization.QuadraticProgram`

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml [hunk 5]
```diff
@@ -0,0 +1,86 @@
+---
+upgrade:
+  - |
+    Shows a warning message if non-printable strings are set to
+    :class:`~qiskit_optimization.problems.QuadraticProgram`
```
Comment (woodsp-ibm) [inline releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:5]:
```suggestion
    :class:`~qiskit_optimization.QuadraticProgram`
```

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    as problem name, variable name, or constraint name.
+  - |
+    If users set an empty variable name ``""`` with
+    :meth:`~qiskit_optimization.problems.QuadraticProgram.binary_var`,
+    :meth:`~qiskit_optimization.problems.QuadraticProgram.integer_var`, and
```
Comment (woodsp-ibm) [inline releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:10]:
```suggestion
    :meth:`~qiskit_optimization.QuadraticProgram.binary_var`,
    :meth:`~qiskit_optimization.QuadraticProgram.integer_var`, and
```

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    :meth:`~qiskit_optimization.problems.QuadraticProgram.continuous_var`,
```
Comment (woodsp-ibm) [inline releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:11]:
```suggestion
    :meth:`~qiskit_optimization.QuadraticProgram.continuous_var`,
```

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        x = qp.binary_var(name="")
+        y = qp.integer_var(name="")
+        z = qp.continuous_var(name="")
+        print(x.name)  # x0
+        print(y.name)  # x1
+        print(z.name)  # x2
+
+features:
+  - |
+    Adds a method :meth:`~qiskit_optimization.problems.QuadraticProgram.prettyprint`
```
Comment (woodsp-ibm) [inline releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:29]:
```suggestion
    Adds a method :meth:`~qiskit_optimization.QuadraticProgram.prettyprint`
```

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    to ``QuadraticProgram`` to generate a pretty-printed string of a ``QuadraticProgram`` object.
+
+    Here is an example of pretty printing.
+
+    .. code-block:: python
+
+        from qiskit_optimization.problems import QuadraticProgram
```
Comment (woodsp-ibm) [inline releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:36]:
```suggestion
        from qiskit_optimization import QuadraticProgram
```

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        y = qp.integer_var(name="")
+        z = qp.continuous_var(name="")
+        print(x.name)  # x0
+        print(y.name)  # x1
+        print(z.name)  # x2
+
+features:
+  - |
+    Adds a method :meth:`~qiskit_optimization.QuadraticProgram.prettyprint`
+    to ``QuadraticProgram`` to generate a pretty-printed string of a ``QuadraticProgram`` object.
```
Comment (woodsp-ibm) [inline releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml:30]:
```suggestion
    to :class:`~qiskit_optimization.QuadraticProgram` to generate a pretty-printed string of the object.
```
With the double single quotes its highlighted and looks like a link but is not, this makes it a link so the user can click on it if they want to go to the class itself.

File: releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml (hunk 3 above)
Comment (woodsp-ibm) [reply releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml]: The references now work - I made one small final change to the text.

Comment (t-imamichi) [reply releasenotes/notes/add-prettyprint-8ec601f41c9cc7d2.yaml]: Thanks!

**Review by woodsp-ibm — APPROVED** [review]

**Pull request merged**